azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +101,242 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+                    "signatureRequired": bool
+                }
+            }
+        }
+
+        If "signatureRequired" isn't provided, the global "signatureRequired" value will be used instead.
+        """
+        extensions = self.__parse_dict(extensions_policy, "extensions")
+        if extensions is None:
+            return
+
+        # Validate "extensions" dict against the schema
+        extensions_schema = _POLICY_SCHEMA["extensionPolicies"]["extensions"]
+        if not isinstance(extensions, dict):
+            raise InvalidPolicyError("invalid type {0} for attribute 'extensions', please change to object."
+                                     .format(type(extensions).__name__))
+
+        # Parse "extensions" and update self._policy with specified attributes
+        parsed_extensions_dict = {}
+        for (extension_name, individual_policy) in extensions.items():
+            self.__validate_schema(individual_policy, extensions_schema["<extensionName>"])
+            if not isinstance(individual_policy, dict):
+                raise InvalidPolicyError("invalid type {0} for attribute '{1}', please change to object."
+                                         .format(type(individual_policy).__name__, extension_name))
+
+            extension_signature_policy = _PolicyEngine.__parse_bool(individual_policy, "signatureRequired", extension_name)
+            if extension_signature_policy is None:
+                extension_signature_policy = self._policy["extensionPolicies"]["signatureRequired"]
+
+            policy_to_add = {
+                "signatureRequired": extension_signature_policy
+            }
+            parsed_extensions_dict[extension_name] = policy_to_add
+
+        # Convert "extensions" to a case-folded dict for case-insensitive lookup
+        case_folded_extensions_dict = _CaseFoldedDict.from_dict(parsed_extensions_dict)
+        self._policy["extensionPolicies"]["extensions"] = case_folded_extensions_dict
+
+    @staticmethod
+    def __validate_schema(policy, schema):
+        """
+        Validate the provided policy against the schema - we only do a shallow check (no recursion into nested dicts).
+        If there is an unrecognized attribute, raise an error.
+        """
+        for key in policy:
+            if key not in schema:
+                raise InvalidPolicyError("attribute '{0}' is not defined in the policy schema. Please refer to the policy documentation "
+                                         "and change or remove this attribute accordingly.".format(key))
+
+    @staticmethod
+    def __parse_bool(policy, key, parent_attribute_name=None):
+        """
+        Raise error if policy[key] is not a boolean. Return none if the key does not exist.
+        For attributes that are used in multiple places (ex: "signatureRequired"), we specify the parent attribute in
+        the error message to avoid ambiguity.
+        """
+        value = policy.get(key)
+        if value is None:
```

Comment:
what if the policy file specifies a null value (like { "signatureRequired": None }) - what is the expected behavior? with this implementation, we don't raise an error and use the default value

Reply:
We spec'd that None/null should be rejected. Updated validation accordingly.